### PR TITLE
[UCDC]borovenskiy/feature/sso

### PR DESCRIPTION
**Pull request by NikolayBorovenskiy**
Youtrack task - https://youtrack.raccoongang.com/issue/UCDC-49.

PR contains changes for integrating Single Sign On for UCDC portal and Edx platform. 
UCDC portal is developed using the Django framework.

**Doc to installation**
https://docs.google.com/document/d/1TpgeIMeFCRRofSO0H082Rm0uSNH4IDBDVFzrYBg6GMQ/edit?usp=sharing

**IMPORTANT**: After a merge of the PR please create a Tag. ucdc-v.0.1.0

### CODE REVIEW

**Review by arsentur — APPROVED** [review]

**Review by Masergik — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -6,7 +6,7 @@ Redirect uri must be **http://<edx_url>/auth/complete/custom-oauth2/**
 
  - Install this client
    ```
-   pip install git+https://github.com/raccoongang/edx-oauth-client.git@hawthorn-master#egg=edx_oauth_client
+   pip install git+https://github.com/raccoongang/edx-oauth-client.git@ucdc-hawthorn-master#egg=edx_oauth_client
```
Comment (idegtiarov) [inline README.md]: Do we propose installing the client from the branch directly? I would prefer to install release pinned to the tag.

File: edx_oauth_client/middleware.py [hunk 2]
```diff
@@ -1,24 +1,46 @@
-import re
-import os.path
+import logging
+from urlparse import urljoin, urlparse
 
+import re
```
Comment (idegtiarov) [inline edx_oauth_client/middleware.py]: `re` is the default python lib so should be added into the first import block.

File: edx_oauth_client/middleware.py [hunk 3]
```diff
@@ -1,24 +1,46 @@
...
 from django.conf import settings
-from django.core.urlresolvers import reverse
 from django.contrib.auth import REDIRECT_FIELD_NAME, logout
 from django.shortcuts import redirect
+from django.urls import reverse
+from social_django.views import auth, NAMESPACE
+
 from edx_oauth_client.backends.generic_oauth_client import GenericOAuthBackend
+from edx_oauth_client.constants import LOCAL_URLS, API_URLS, OAUTH_PROCESS_URLS
```
Comment (idegtiarov) [inline edx_oauth_client/middleware.py]: Please reorder imported params in alphabetical order.

File: edx_oauth_client/middleware.py [hunk 4]
```diff
@@ -1,24 +1,46 @@
...
+
+
+    def process_response(self, request, response):
+        """
+        Delete cross-domain cookie of SSO flow to accomplish logout.
+        """
+        if not request.session.get(self.cookie_name):
+            response.set_cookie(
+                self.cookie_name,
+                domain=self.cookie_domain,
```
Comment (idegtiarov) [inline edx_oauth_client/middleware.py:36]: I would add validation that `self.cookies_domain` is not None.

File: edx_oauth_client/middleware.py [hunk 5]
```diff
@@ -46,14 +68,21 @@ def process_request(self, request):
...
+
         if not is_same_user and is_auth:
             logout(request)
 
         if (auth_cookie and not is_continue and (not is_auth or not is_same_user)) or \
                 ('force_auth' in request.session and request.session.pop('force_auth')):
             query_dict = request.GET.copy()
             query_dict[REDIRECT_FIELD_NAME] = current_url
-            query_dict['auth_entry'] = 'login'
+            query_dict["auth_entry"] = "login"
```
Comment (idegtiarov) [inline edx_oauth_client/middleware.py:83]: Why was this change added?

File: edx_oauth_client/middleware.py [hunk 6]
```diff
@@ -69,69 +98,32 @@ def process_request(self, request):
...
-                provider_url = PROVIDER_URL.rstrip("/") + "/"
-                return redirect("%s%s" % (provider_url, current_url))
-
-        if not is_auth and start_url not in auth_process_urls and start_url not in api_urls:
+            debug_handle_local_urls = ("debug", settings.STATIC_URL, settings.MEDIA_URL)
+            available_urls += debug_handle_local_urls
+
+        if request.user.is_authenticated():
+            if dashboard_url_path and request.path.strip("/") in reverse("dashboard"):
+                return redirect(urljoin(provider_url, dashboard_url_path))
```
Comment (idegtiarov) [inline edx_oauth_client/middleware.py:116]: I would add some logs to cover the case url parameters were not provided correctly

File: edx_oauth_client/middleware.py [hunk 7]
```diff
@@ -69,69 +98,32 @@ def process_request(self, request):
...
+
+            if courses_list_url_path and request.path.strip("/") in reverse("courses"):
+                return redirect(urljoin(provider_url, courses_list_url_path))
+
+            if (
+                    user_account_url_path
```
Comment (idegtiarov) [inline edx_oauth_client/middleware.py]: too big indentation, please take a look on the open edx python code style https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python-guidelines.html#breaking-long-lines

**Review by idegtiarov — CHANGES_REQUESTED** [review]
Some minor nits and questions are inline. Please take a look.

File: edx_oauth_client/middleware.py (hunk 2 above)
Comment (NikolayBorovenskiy) [reply edx_oauth_client/middleware.py]: Was optimized by my IDE and here I totally agree, that optimization was with claims. Fixed the order.

File: edx_oauth_client/middleware.py (hunk 3 above)
Comment (NikolayBorovenskiy) [reply edx_oauth_client/middleware.py]: done

File: edx_oauth_client/middleware.py (hunk 5 above)
Comment (NikolayBorovenskiy) [reply edx_oauth_client/middleware.py:83]: As far as I know, in python double quotes are preferable to single quotes. I try to lead to this principle. If you do not mind, then in this module I will bring the quotes to a general view(double). Open to discussion.
Pushed a fix.

File: edx_oauth_client/middleware.py (hunk 7 above)
Comment (NikolayBorovenskiy) [reply edx_oauth_client/middleware.py]: yeah. Added a new version of the complex `if`

File: edx_oauth_client/middleware.py (hunk 6 above)
Comment (NikolayBorovenskiy) [reply edx_oauth_client/middleware.py:116]: I do not think this should be done for several reasons. 
1. An incorrect redirect will not do anything bad. Either nothing will happen or it will simply be in the wrong place where we expect.
2. These are subtleties of tuning and debugging. And it is our responsibility to make the configuration correct. Additionally, we make a document where all this is described.
3. It is not clear how to validate and what to be afraid of.

File: README.md (hunk 1 above)
Comment (NikolayBorovenskiy) [reply README.md]: Good point thnx. Pushed a fix.

**Comment by NikolayBorovenskiy** [comment]
IMPORTANT: After a merge of the PR please create a Tag. ucdc-v.0.1.0

File: edx_oauth_client/middleware.py (hunk 6 above)
Comment (idegtiarov) [reply edx_oauth_client/middleware.py:116]: agree, tnx for the verbose explanation. 0))

**Review by idegtiarov — APPROVED** [review]
